src/rpc/core_rpc_server.cpp

```diff
@@ -2907,5 +2907,44 @@ namespace cryptonote
     res.status = CORE_RPC_STATUS_OK;
     return true;
   }
+  //------------------------------------------------------------------------------------------------------------------------------
+  bool core_rpc_server::on_get_checkpoints(const COMMAND_RPC_GET_CHECKPOINTS::request& req, COMMAND_RPC_GET_CHECKPOINTS::response& res, epee::json_rpc::error& error_resp, const connection_context *ctx)
+  {
+    bool unused_ = false;
+    if (use_bootstrap_daemon_if_necessary<COMMAND_RPC_GET_CHECKPOINTS>(invoke_http_mode::JON_RPC, "get_checkpoints", req, res, unused_))
+      return true;
+
+    size_t num_checkpoints = req.num_checkpoints_to_query;
+    if (ctx)
+      num_checkpoints = std::min((size_t)256, num_checkpoints);
+
+    Blockchain const &blockchain = m_core.get_blockchain_storage();
+    BlockchainDB const &db       = blockchain.get_db();
+
+    if (req.heights.size())
+    {
+      checkpoint_t checkpoint;
+      for (size_t i = 0; i < num_checkpoints; i++)
+      {
+        uint64_t height = req.heights[i];
+        if (db.get_block_checkpoint(height, checkpoint))
+          res.checkpoints.push_back(checkpoint);
+      }
+    }
+    else
+    {
+      checkpoint_t top_checkpoint;
+      if (db.get_top_checkpoint(top_checkpoint))
+      {
+        uint64_t start = top_checkpoint.height < num_checkpoints
+                             ? 0
+                             : top_checkpoint.height - num_checkpoints;
+        res.checkpoints = db.get_checkpoints_range(top_checkpoint.height, 0, num_checkpoints);
```

Comment:
I don't understand this logic.

This code in this path is meant to return the last N checkpoints.  So let's say I call it with N=3 (i.e. I pass `num_checkpoints_to_query` set to 3 in the JSON RPC call).

First my `N` value gets bumped up to 256 for some reason (I suppose so that the N=60 default ends up being approximately right assuming a checkpoint every 4 blocks).

So then if the top checkpoint height is, say, 200000 then I ask the database for 256 checkpoints starting at 199744.  So I get back... well, some number of checkpoints that is almost certainly bigger than 3.

Or another example: if I ask the RPC for the last N=100 checkpoints, the database is going to return... well, still just exactly what it would return if I asked for N=3, but in this case it is almost certainly less than 100.

Indeed unless I start passing it values of N greater than 256 it seems like I will always get exactly the same vector of checkpoints out of this function, but I am almost never going to get back the `num_checkpoints` that I asked for.

Reply:
Am I missing something, or did you confuse the N = std::min(256, N) with a std::max()? That start variable is unused though, so I'll remove.

Reply:
Blech, yes, I did.  :(

But in that case it seems like I am always going to get fewer results than I asked for because:

`top_checkpoint.height - num_checkpoints` is basically adding different units together.

Reply:
Yeah that was an mistake that propagated itself throughout the whole PR so gonna be a bit of rewriting to do.

Reply:
(But I think this basically feeds into the other point about how `num_checkpoints` gets interpreted).